src/main/java/com/xiaomi/infra/pegasus/client/PegasusAdminClient.java

```diff
@@ -0,0 +1,189 @@
+    implements PegasusAdminClientInterface {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PegasusClient.class);
+  private static final String APP_TYPE = "pegasus";
+
+  private Meta meta;
+
+  @Override
+  public String clientType() {
+    return "Admin";
+  }
+
+  public PegasusAdminClient(Properties properties) throws PException {
+    this(ClientOptions.create(properties));
+  }
+
+  public PegasusAdminClient(String configPath) throws PException {
+    this(ClientOptions.create(configPath));
+  }
+
+  public PegasusAdminClient(ClientOptions options) throws PException {
+    super(options);
+    this.meta = this.cluster.getMeta();
+    LOGGER.info(
+        "Create PegasusToolsClient Instance By ClientOptions : {}", this.clientOptions.toString());
+  }
+
+  @Override
+  public void createApp(String appName, int partitionCount, int replicaCount, long timeoutMs)
+      throws PException {
+    if (partitionCount < 1) {
+      throw new PException(
+          new IllegalArgumentException("createApp  failed: partitionCount should >= 1!"));
+    }
+
+    if (replicaCount < 1) {
+      throw new PException(
+          new IllegalArgumentException("createApp  failed: replicaCount should >= 1!"));
+    }
+
+    int i = 0;
+    for (; i < appName.length(); i++) {
+      char c = appName.charAt(i);
+      if (!((c >= 'a' && c <= 'z')
+          || (c >= 'A' && c <= 'Z')
+          || (c >= '0' && c <= '9')
+          || c == '_'
+          || c == '.'
+          || c == ':')) {
+        break;
+      }
+    }
+
+    if (appName.isEmpty() || i < appName.length()) {
+      throw new PException(
+          new IllegalArgumentException(
+              String.format("createApp  failed: invalid appName: %s", appName)));
+    }
+
+    if (timeoutMs <= 0) {
+      throw new PException(
+          new IllegalArgumentException(
+              String.format("createApp  failed: invalid timeoutMs: %d", timeoutMs)));
+    }
+
+    long startTime = System.currentTimeMillis();
+
+    create_app_options options = new create_app_options();
+    options.setPartition_count(partitionCount);
+    options.setReplica_count(replicaCount);
+    options.setSuccess_if_exist(true);
+    options.setApp_type(APP_TYPE);
+    options.setEnvs(new HashMap<>());
+    options.setIs_stateful(true);
+
+    configuration_create_app_request request = new configuration_create_app_request();
+    request.setApp_name(appName);
+    request.setOptions(options);
+
+    create_app_operator app_operator = new create_app_operator(appName, request);
+    error_code.error_types error = this.meta.operate(app_operator, timeoutMs);
+    if (error != error_code.error_types.ERR_OK) {
+      throw new PException(
+          String.format(
+              "Create app:%s failed, partitionCount: %d, replicaCount: %s, error:%s.",
+              appName, partitionCount, replicaCount, error.toString()));
+    }
+
+    long endCreateAppRpcTime = System.currentTimeMillis();
```

Comment:
You can calc a deadline ts at the beginning, then compare it with currentTimeMillis to check whether timeout. The code would be more simple.